src/promql/src/extension_plan/union_distinct_on.rs

```diff
@@ -0,0 +1,576 @@
+                let right = ready!(fut.as_mut().poll(cx))?;
+                self.right = HashedDataFut::Ready(right);
+                let HashedDataFut::Ready(right_ref) = &mut self.right else {
+                    unreachable!()
+                };
+                right_ref
+            }
+            HashedDataFut::Ready(ref mut right) => right,
+            HashedDataFut::Empty => return Poll::Ready(None),
+        };
+
+        // poll left and probe with right
+        let next_left = ready!(self.left.poll_next_unpin(cx));
+        match next_left {
+            Some(Ok(left)) => {
+                // observe left batch and return it
+                right.update_map(&left)?;
+                Poll::Ready(Some(Ok(left)))
+            }
+            Some(Err(e)) => Poll::Ready(Some(Err(e))),
+            None => {
+                // left stream is exhausted, so we can send the right part
+                let right = std::mem::replace(&mut self.right, HashedDataFut::Empty);
+                let HashedDataFut::Ready(data) = right else {
+                    unreachable!()
+                };
+                Poll::Ready(Some(data.finish()))
+            }
+        }
+    }
+}
+
+impl RecordBatchStream for UnionDistinctOnStream {
+    fn schema(&self) -> SchemaRef {
+        self.output_schema.clone()
+    }
+}
+
+impl Stream for UnionDistinctOnStream {
+    type Item = DataFusionResult<RecordBatch>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.poll_impl(cx)
+    }
+}
+
+/// Simple future state for [HashedData]
+enum HashedDataFut {
+    /// The result is not ready
+    Pending(BoxFuture<'static, DataFusionResult<HashedData>>),
+    /// The result is ready
+    Ready(HashedData),
+    /// The result is taken
+    Empty,
+}
+
+/// ALL input batches and its hash table
+struct HashedData {
+    // TODO(ruihang): use `JoinHashMap` instead after upgrading to DF 34.0
+    /// Hash table for all input batches. The key is hash value, and the value
+    /// is the index of `bathc`.
+    hash_map: HashMap<u64, usize>,
+    /// Output batch.
+    batch: RecordBatch,
+    /// The indices of the columns to be hashed.
+    hash_key_indices: Vec<usize>,
+    random_state: RandomState,
+}
+
+impl HashedData {
+    pub async fn new(
+        input: SendableRecordBatchStream,
+        random_state: RandomState,
+        hash_key_indices: Vec<usize>,
+    ) -> DataFusionResult<Self> {
+        // Collect all batches from the input stream
+        let initial = (Vec::new(), 0);
+        let (batches, _num_rows) = input
+            .try_fold(initial, |mut acc, batch| async {
+                // Update rowcount
+                acc.1 += batch.num_rows();
+                // Push batch to output
+                acc.0.push(batch);
+                Ok(acc)
+            })
+            .await?;
+
+        // Create hash for each batch
+        let mut hash_map = HashMap::default();
```

Comment:
Question: does the execution plan or output stream contain any statistics about the output that can help us allocate the proper size of `HashMap`? Try to avoid some memory allocations.

Reply:
We don't have such a thing for now. The incoming stream might be complex and we cannot use file statistics here either.